server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -30,16 +31,27 @@ public static DocumentMapper createEmpty(MapperService mapperService) {
         );
         MetadataFieldMapper[] metadata = mapperService.getMetadataMappers().values().toArray(new MetadataFieldMapper[0]);
         Mapping mapping = new Mapping(root, metadata, null);
-        return new DocumentMapper(mapperService.documentParser(), mapping, mapping.toCompressedXContent());
+        return new DocumentMapper(mapperService.documentParser(), mapping, mapping.toCompressedXContent(), IndexVersion.current());
     }
 
-    DocumentMapper(DocumentParser documentParser, Mapping mapping, CompressedXContent source) {
+    DocumentMapper(DocumentParser documentParser, Mapping mapping, CompressedXContent source, IndexVersion version) {
         this.documentParser = documentParser;
         this.type = mapping.getRoot().name();
         this.mappingLookup = MappingLookup.fromMapping(mapping);
-        this.mappingSource = source;
-        assert mapping.toCompressedXContent().equals(source)
-            : "provided source [" + source + "] differs from mapping [" + mapping.toCompressedXContent() + "]";
+        if (sourceMapper().isSynthetic()
+            && source.string().contains("\"_source\":{\"mode\":\"synthetic\"}") == false
+            && version.onOrBefore(IndexVersion.V_8_10_0)) {
+            /*
+             * Indexes built at v.8.7 were missing an explicit entry for synthetic_source.
+             * This got restored in v.8.9 (and patched in v.8.8) to avoid confusion. The change is only restricted to
+             * mapping printout, it has no functional effect as the synthetic source already applies.
+             */
+            this.mappingSource = mapping.toCompressedXContent();
```

Comment:
By doing `this.mappingSource = mapping.toCompressedXContent();` we are changing the mapping source to contain the missing `_source: mode: synthetic` right?

If we would do `this.mappingSource = source;` here, then the that would remain to be missing?

Reply:
Correct. I was thinking of storing a source that matches this output to avoid surprises further down the processing pipeline, but up to you.